Code/Framework/AzCore/AzCore/Module/Environment.h

```diff
@@ -242,41 +256,29 @@ namespace AZ
                 memset(&m_value, 0, sizeof(T));
             }
 
-            template <class... Args>
+            template<class... Args>
             void ConstructImpl(const AZStd::false_type& /* AZStd::has_trivial_constructor<T> */, Args&&... args)
             {
                 // Construction of non-trivial types is left up to the type's constructor.
                 new(&m_value) T(AZStd::forward<Args>(args)...);
             }
-
-            void DestructImpl(const AZStd::true_type& /* AZStd::is_trivially_destructible<T> */)
+            static void DestructDispatchNoLock(EnvironmentVariableHolderBase *base, DestroyTarget selfDestruct)
             {
-                // do nothing
-            }
-
-            void DestructImpl(const AZStd::false_type& /* AZStd::is_trivially_destructible<T> */)
-            {
-                reinterpret_cast<T*>(&m_value)->~T();
-            }
-
-            // Assumes the lock is already held
-            void UnregisterAndDestruct()
-            {
-                // if the environment that created us is gone the owner can be null
-                // which means (assuming intermodule allocator) that the variable is still alive
-                // but can't be found as it's not part of any environment.
-                if (m_environmentOwner)
+                auto *self = reinterpret_cast<EnvironmentVariableHolder *>(base);
+                if (selfDestruct == DestroyTarget::Self)
                 {
-                    m_environmentOwner->RemoveVariable(m_guid);
-                    m_environmentOwner = nullptr;
+                    self->~EnvironmentVariableHolder();
+                    return;
                 }
 
-                if (m_isConstructed)
+                AZ_Assert(self->m_isConstructed, "Variable is not constructed. Please check your logic and guard if needed!");
+                self->m_isConstructed = false;
+                self->m_moduleOwner = nullptr;
+                if constexpr(AZStd::is_trivially_destructible_v<T>)
                 {
-                    DestructNoLock();
+                    reinterpret_cast<T*>(&self->m_value)->~T();
                 }
```

Comment:
Should this be `if constexpr(!AZStd::is_trivially_destructible_v<T>)` because a trivial destructor doesn't require an explicit call and memory can just be released: https://en.cppreference.com/w/cpp/language/destructor#Trivial_destructor "Objects with trivial destructors don't require a delete-expression and may be disposed of by simply deallocating their storage."

Reply:
That is what is test failure!

The memory leak has been found.

Reply:
I'm almost sure I've fixed that :smile: must've been the previous attempt...

Reply:
Yup, I've fixed it in b3419e9  but the PR it was a part of was closed.

Now it's fixed here as well.